src/mlpack/methods/ann/loss_functions/cosine_embedding_loss_impl.hpp

```diff
@@ -0,0 +1,142 @@
+/**
+ * @file cosine_embedding_loss_impl.hpp
+ * @author Kartik Dutt
+ *
+ * Implementation of the Cosine Embedding loss function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_COSINE_EMBEDDING_IMPL_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_COSINE_EMBEDDING_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "cosine_embedding_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+CosineEmbeddingLoss<InputDataType, OutputDataType>::CosineEmbeddingLoss(
+    const double margin, const bool takeMean):
+    margin(margin), takeMean(takeMean)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<
+    typename FirstTensor,
+    typename SecondTensor,
+    typename ThirdTensor
+>
+double CosineEmbeddingLoss<InputDataType, OutputDataType>::Forward(
+    const FirstTensor&& x1,
+    const SecondTensor&& x2,
+    const ThirdTensor&& y)
+{
+  const size_t cols = x1.n_cols;
+  const size_t batchSize = x1.n_elem / cols;
+  if (x1.n_rows != x2.n_rows || x1.n_cols != x2.n_cols ||
+      x1.n_elem != x2.n_elem)
```

Comment:
I might be missing something, but how about using ```arma::size(x1)!=arma::size(x2)```.

Reply:
Already switched, thanks anyway.